pkg/kv/kvserver/rangefeed/scheduled_processor.go

```diff
@@ -574,19 +575,37 @@ func (p *ScheduledProcessor) Filter() *Filter {
 // is guaranteed that only single request is modifying processor at any given
 // time. It is advisable to use provided processor reference for operations
 // rather than using one within closure itself.
-// If request can't be queued or processor stoppedC is closed then default
-// value is returned.
+//
+// If the processor is stopped concurrently with the request queueing, it may or
+// may not be processed. If the request is ever processed, its return value is
+// guaranteed to be returned here. Otherwise, the zero value is returned and the
+// request is never processed.
 func runRequest[T interface{}](
 	p *ScheduledProcessor, f func(ctx context.Context, p *ScheduledProcessor) T,
 ) (r T) {
 	result := make(chan T, 1)
 	p.enqueueRequest(func(ctx context.Context) {
 		result <- f(ctx, p)
+		// Assert that we never process requests after stoppedC is closed. This is
+		// necessary to coordinate catchup iter ownership and avoid double-closing.
+		if buildutil.CrdbTestBuild {
+			select {
+			case <-p.stoppedC:
+				log.Fatalf(ctx, "processing request on stopped processor")
+			default:
+			}
+		}
```

Comment:
What if `stoppedC` was closed right after `f` was invoked and written to `result`?

Is the invariant that `stoppedC` is always closed in between processing the enqueued requests? So it should be either closed before this callback (then the callback won't be called), or after it; but not in the middle of the callback?

In this sense, is it true that this invariant check can equivalently be either before or after line 588? Or does it have to come after line 588?

Reply:
`f()` has exclusive access to `p` for the duration of the call, as documented on `runRequest()`. It follows that the caller (this function) has exclusive access to it.

For intuition, consider that `ScheduledProcessor.process()` is single-threaded, and that both `processRequest()` and `processStop()` are called from it -- the latter closes `stoppedC`.